From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sleigh-InSPECtor <sleighinspector@outlook.com>
Date: Thu, 16 May 2024 13:37:25 +0930
Subject: [PATCH] 6530: AArch32: sha1su1.32 had destructive left shifts on
 32-bit values before zext

AArch32: fixed sha1su1.32

* fixed destructive left shifts on 32-bit values before zext
---
 Ghidra/Processors/ARM/data/languages/ARMneon.sinc | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/Ghidra/Processors/ARM/data/languages/ARMneon.sinc b/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
index f558f7f921..d047b3a9d1 100644
--- a/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
+++ b/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
@@ -637,15 +637,15 @@ define pcodeop SHA1HashUpdateParity;
 	local X = Qd;
 	local Y = Qm;
 	local Tm = X ^ (Y >> 32);
-	local t0:4 = Tm(0);
-	local t1:4 = Tm(4);
-	local t2:4 = Tm(8);
-	local t3:4 = Tm(12);
+	local t0:4 = Tm[0, 32];
+	local t1:4 = Tm[32, 32];
+	local t2:4 = Tm[64, 32];
+	local t3:4 = Tm[96, 32];
 	local W0:4 = (t0 << 1 | t0 >> 31);
 	local W1:4 = (t1 << 1 | t1 >> 31);
 	local W2:4 = (t2 << 1 | t2 >> 31);
 	local W3:4 = (t3 << 1 | t3 >> 31) ^ (t0 << 2 | t0 >> 30);
-	Qd = zext(W3 << 96) | zext(W2 << 64) | zext(W1 << 32) | zext(W0);
+	Qd = (zext(W3) << 96) | (zext(W2) << 64) | (zext(W1) << 32) | zext(W0);
 }
 
 #######
-- 
2.45.0

